### PR TITLE
refactor, better api, queue and stacktracing

### PR DESCRIPTION
**Pull request by hallas**

### CODE REVIEW

**Comment by hallas** [comment]
@roncohen 

I don't think the code will get any cleaner than this, and as far as I can tell this covers the original functionality and more. It's DRY and I like the API. There are a number of possible improvements already, such as using constants for `level` parameter, possibly other logging and of course the revision flag parsed via env variables, but that should be really easy.

It would be awesome if you helped me on the test cases in terms of the Opbeat backend and I guess also we have to create a coverage report.

I'd like to merge this in, and as you can see in the travis configuration, we're only testing for 1.3+ because of the `http` library dependency.

File: opbeat.go [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+
+func CaptureMessage(message string, l Level) error {
+	return DefaultOpbeat.CaptureMessage(message, l)
+}
+
+func Wait() {
+	DefaultOpbeat.Wait()
+}
+
+func Start() {
```
Comment (roncohen) [inline opbeat.go]: What is the purpose of exposing the `Start` commands? seems like it's always called

File: opbeat.go [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+// Waits for all packets to send.
+func (opbeat *Opbeat) Wait() {
+	opbeat.wait.Wait()
+}
+
+// Starts a goroutine which listens on the main channel and sends any packets it
+// receives to Opbeat. The goroutine will exit if the client's `.Close` function
+// is called.
+func (opbeat *Opbeat) Start() {
+	opbeat.packets = make(chan *packet)
```
Comment (roncohen) [inline opbeat.go]: This channel should probably be buffered so callers of `CaptureX` won't block if another request to Opbeat is in progress.

**Comment by roncohen** [comment]
Great work @hallas !

I've added a couple of notes. Nothing major. I also think you forgot to add the new doc.go ?

File: opbeat.go (hunk 1 above)
Comment (hallas) [inline opbeat.go]: I don't think there is any real use-case but technically you can Close it and then Start it again.

Comment (hallas) [inline opbeat.go]: If you want to you can "Start" twice to increase concurrency.

File: opbeat.go (hunk 2 above)
Comment (hallas) [inline opbeat.go]: How large a buffer are you thinking here?

**Comment by hallas** [comment]
It's in the opbeat.go file - I don't think it has to be in a doc.go file.

File: opbeat.go (hunk 1 above)
Comment (roncohen) [inline opbeat.go]: Starting twice won't work currently, because `opbeat.packets` gets replaced for every call.

Comment (roncohen) [inline opbeat.go]: I vote in favor of removing `Start` until we find a usecase.

**Comment by roncohen** [comment]
Ok. I was looking for an example of how to pass additional info using `Extra`, but couldn't find it - because it's no longer possible ;)

I think there is still a legitimate use case for sending custom data in `Extra`. Can we add it back?

File: opbeat.go (hunk 2 above)
Comment (roncohen) [inline opbeat.go]: Good question. Let's do 20

**Comment by hallas** [comment]
Yup. I think we can use the params, the same way we do it for the http information.

File: opbeat.go (hunk 1 above)
Comment (hallas) [inline opbeat.go]: I agree, and yes, you're right about the channel

**Comment by hallas** [comment]
@roncohen

In terms of API, should you be able to pass extra to both CaptureError and CaptureMessage? And are you ok with it being through an "options" map?

**Comment by roncohen** [comment]
+1 for options map

**Comment by hallas** [comment]
@roncohen så er der options -> "http" og "extra" er understøtet nu

**Comment by roncohen** [comment]
Awesome Hallas. What do you think about that extra nil versus a method called `CaptureErrorWithOptions` ?

**Comment by hallas** [comment]
Updated with types for options and user field on the packet including a test to cover it

**Comment by hallas** [comment]
84.8% coverage

```
$ go test -coverprofile=coverage.out
$ go tool cover -html=coverage.out
```

looks good enough for me ;-)

**Comment by hallas** [comment]
Hoping you have fun in Sweden! I think we should go ahead and add more convenient methods. But let us compile a list?

```
- CaptureErrorWithOptions
- CaptureErrorWithExtra
```

et cetera

**Comment by roncohen** [comment]
It's a tough balance. We don't want to have a very big API. I'm planing to take a closer look over the weekend.  Good job on the coverage btw!

**Comment by roncohen** [comment]
It seems to me like what we're doing now is fine (everything takes optional `Option` which can be `nil`). No need to add `*WithOptions` or `*WithExtra`. I'll get the server side stuff required for `rev` deployed asap, and then we can merge this. :+1:

**Comment by roncohen** [comment]
By the way, i think we should still have an example that shows advanced usage. This was from previously:

------

Example		
	A complete example could look like this:		
		
	config := ClientConfig{		
		OrganizationId: OrgId,		
		AppId:          AppId,		
		SecretToken:    SecretToken,		
		Logger:         log.New(os.Stderr, "OPBEAT ", log.LstdFlags),		
	}		
		
	client, _ := NewClient(&config)		
	_, err := client.CaptureErrorWithOptions(		
		errors.New("Waaat!"),		
		&EventOptions{		
			User: &User{		
				Id:              "99",		
				Username:        "roncohen",		
				Email:           "ron@opbeat.com",		
				IsAuthenticated: true,		
			},		
			Http: NewHttpFromRequest(req),		
			Extra: &map[string]interface{}{"hello": "world"},		
		},		
	)
```

**Comment by hallas** [comment]
Kan være at når vi er færdige og klar, at vi kan demo skidtet sammen til en cph gophers meetup? og/eller andre meetups.

**Comment by roncohen** [comment]
This is interesting: http://changelog.ca/log/2015/01/30/golang

Though, you loose the documentation value and clarity of having strictly defined parameters to methods.

**Comment by hallas** [comment]
Hmm. Not into that because it doesn't seem very "Go"-like ?

**Comment by roncohen** [comment]
Yeah, i agree. Let's leave it. Time will tell if we've chosen the right interface ;)

File: opbeat.go [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+	if err != nil {
+		return err
+	}
+
+	body, err = ioutil.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+
+	res.Body.Close()
```
Comment (roncohen) [inline opbeat.go]: @birkirb recommends moving this up to L322 and using `defer` to make sure we close even when we get an error :+1:

Comment (hallas) [inline opbeat.go]: :gem: Of course!

**Comment by birkirb** [comment]
You guys don't use GoLint?  It's got some good suggestions on how to write idiomatic Go.  
Also these guidelines buy the Google team are pretty good:  https://code.google.com/p/go-wiki/wiki/CodeReviewComments but I think you are pretty much there already.

**Comment by hallas** [comment]
We run go format on the code that's it, is GoLint a tool or a guideline?

**Comment by birkirb** [comment]
Both. :)
https://github.com/golang/lint

**Comment by hallas** [comment]
thanks @birkirb 

I'll apply the necessary changes

**Comment by hallas** [comment]
@roncohen @birkirb :dancer: :dancers: :ghost: 

Take a look fellas :-)

**Comment by hallas** [comment]
@roncohen in terms of the advanced example, you want to add it in the comments right?

**Comment by hallas** [comment]
:fireworks:

**Pull request merged**